ProjFS.Mac/PrjFSLib/PrjFSUser.cpp

```diff
@@ -11,13 +11,51 @@ struct DarwinVersion
     unsigned long major, minor, revision;
 };
 
+struct PrjFSService_WatchContext
+{
+    std::function<void(io_service_t, io_connect_t, PrjFSService_WatchContext*)> discoveryCallback;
+    io_iterator_t notificationIterator;
+    PrjFSServiceUserClientType clientType;
+};
+
+
 typedef decltype(IODataQueueDequeue)* ioDataQueueDequeueFunctionPtr;
 static ioDataQueueDequeueFunctionPtr ioDataQueueDequeueFunction = nullptr;
 typedef decltype(IODataQueuePeek)* ioDataQueuePeekFunctionPtr;
 static ioDataQueuePeekFunctionPtr ioDataQueuePeekFunction = nullptr;
 
 static void InitDataQueueFunctions();
 
+bool PrjFSService_ValidateVersion(io_service_t prjfsService)
+{
+    CFTypeRef kextVersionObj = IORegistryEntryCreateCFProperty(prjfsService, CFSTR(PrjFSKextVersionKey), kCFAllocatorDefault, 0);
+    CFStringRef kextVersionString;
+    if (nullptr == kextVersionObj || CFStringGetTypeID() != CFGetTypeID(kextVersionObj))
+    {
+        std::cerr << "PrjFS kernel service does not advertise valid kext version property\n";
+        goto CleanupAndFail;
+    }
+    
+    kextVersionString = static_cast<CFStringRef>(kextVersionObj);
+    if (kCFCompareEqualTo != CFStringCompare(kextVersionString, CFSTR(PrjFSKextVersion), 0 /* options: case sensitive */))
```

Comment:
Regarding case sensitivity: I guess so, though I'm not sure what the point would be. At the moment, the version is [hardcoded as `"0.1"`](https://github.com/Microsoft/VFSForGit/blob/982bdc5697f6c7c04809dc06b216b30db5018755/ProjFS.Mac/PrjFSKext/public/PrjFSCommon.h#L14) so there aren't any letters to treat as case sensitive. We should definitely change the versioning, preferably before the first release, but that's kind of out of scope in this PR. The version checking code itself hasn't changed, I just moved it to a helper function.

Reply:
@pmj sorry my mistake here, I misread `options: case sensitive`.  The actual question was wondering if there was a named constant we could use instead of 0, and it looks like there's not.

Please disregard.

Reply:
`CFString`: the reason I used `CFStringCompare` is simply because the code is simpler and more convenient in this context. `IORegistryEntryCreateCFProperty` will always produce Core Foundation objects, so it's a question of embracing that or converting to something else first. I'd recommend a case-by-case assessment, and I reckon in this case the CF approach wins. :-)

Reply:
Thanks that makes sense, I agree the CF approach is clean here :)

Reply:
> @pmj sorry my mistake here, I misread `options: case sensitive`. The actual question was wondering if there was a named constant we could use instead of 0, and it looks like there's not.

If you can think of a less ambiguous way to phrase it, I'm happy to change it. But yeah, the point of the comment is to explain the magic 0.